File: LR2/CPP/SimpleIterations/simple_iterations.cpp
#include "simple_iterations.hpp"


bool CheckConvergence(const std::vector<std::vector<double>> &main_coefficients)
{
    for (std::size_t row = 0; row < main_coefficients.size(); ++row)
    {
        const auto main_absolute{std::abs(main_coefficients.at(row).at(row))};
        std::size_t counter{0};
        const auto rest_sum_absolute{std::accumulate(
                std::begin(main_coefficients.at(row)),
                std::end(main_coefficients.at(row)),
                0.00,
                [&](const double response, const double current) -> double {
                    return counter++ != row ? response + std::abs(current) : response;
                })};

        if (main_absolute < rest_sum_absolute)
            return false;
    }

    return true;
}


std::vector<std::vector<double>> ExpressMainVariables(const std::vector<std::vector<double>> &main_coefficients,
                                                      const std::vector<double> &free_coefficients)
{
    std::vector<std::vector<double>> response{};
    response.resize(main_coefficients.size());

    for (std::size_t row = 0; row < main_coefficients.size(); ++row)
    {
        const auto full_width{main_coefficients.at(row).size() + 1};
        const auto current_variable_ratio{main_coefficients.at(row).at(row)};
        response.at(row).resize(full_width);

        response.at(row).at(0) = free_coefficients.at(row) / current_variable_ratio;

        for (std::size_t col = 1; col < full_width; ++col)
            response.at(row).at(col) =
                    col - 1 != row
                    ? -1 * main_coefficients.at(row).at(col - 1) / current_variable_ratio
                    : 0;
    }

    return response;
}


double GetError(const std::vector<double> &current_variables_set,
                const std::vector<double> &previous_variables_set)
{
    double error{0.00};
    for (std::size_t counter = 0; counter < current_variables_set.size(); ++counter)
        error = std::max(error, std::abs(current_variables_set.at(counter) - previous_variables_set.at(counter)));
    return error;
}


std::pair<std::vector<double>, std::size_t>
SolveBySimpleIterations(const std::vector<std::vector<double>> &main_coefficients,
                        const std::vector<double> &free_coefficients,
                        const double epsilon,
                        const std::vector<double> &initial_values,
                        const SolvingType &type)
{
    if (!CheckConvergence(main_coefficients))
        throw std::invalid_argument("System of linear algebraic equations does not converge");

    const auto check_error([](const std::vector<double> &current_variables_set,
                              const std::vector<double> &previous_variables_set,
                              const double epsilon) -> bool {
        return GetError(current_variables_set, previous_variables_set) <= std::abs(epsilon);
    });

    const auto copy_vectors([](const std::vector<double> &from, std::vector<double> &to) -> void {
        for (std::size_t counter = 0; counter < from.size(); ++counter)
            to.at(counter) = from.at(counter);
    });

    const auto count_variable_value([](const std::vector<double> &variable_expression,
                                       const std::vector<double> &variables_set) -> double {
        double response{variable_expression.at(0)};
        for (std::size_t counter = 1; counter < variable_expression.size(); ++counter)
            response += variable_expression.at(counter) * variables_set.at(counter - 1);
        return response;
    });

    auto previous_iteration{initial_values}; // предыдущая i-я итерация
    auto current_iteration{initial_values}; // текущая i+1-я итерация
    const auto variables_expressions{
            ExpressMainVariables(main_coefficients, free_coefficients)}; // выраженные переменные с главной диагонали

    std::size_t iterations_count{0};
    do
    {
        ++iterations_count;
        copy_vectors(current_iteration, previous_iteration);
        for (std::size_t counter = 0; const auto &variable_expression : variables_expressions)
        {
            current_iteration.at(counter) = count_variable_value(
                    variable_expression,
                    type == SolvingType::kSimpleIterations ? previous_iteration : current_iteration
            );
            ++counter;
        }
    } while (!check_error(current_iteration, previous_iteration, epsilon));

    return {current_iteration, iterations_count};
}

File: LR2/Py/simple_iterations.py
import data
from utils import get_copy
from enum import Enum


class SimpleIterationsSolvingType(Enum):
    DEFAULT = 'DEFAULT',
    SEIDEL = 'SEIDEL'


def express_main_diagonal_variables(main_coefficients: list, free_coefficients: list) -> list:
    response = get_copy(main_coefficients)
    for row in response:
        row.append(0)

    for row in range(0, len(main_coefficients)):
        full_width = len(main_coefficients[row]) + 1
        current_variable_ratio = main_coefficients[row][row] + 0

        response[row][0] = free_coefficients[row] / current_variable_ratio
        for col in range(1, full_width):
            response[row][col] = 0 if col - 1 == row \
                else (-1) * main_coefficients[row][col - 1] / current_variable_ratio

    return response


# Подсчёт значения переменной с текущими значениями переменных с предыдущей итерации
# Например, x2 = 6 + 0.3 * x1 + 0 * x2 + 3 * x3 => при наличии с прошлой итерации переменных x1 .. x3 считаю новое значение для x3
def count_variable_value(variable_expression: list, with_variable_set: list) -> float:
    response = variable_expression[0]
    for counter in range(1, len(variable_expression)):
        response += with_variable_set[counter - 1] * variable_expression[counter]
    return response


def count_all_variables_values(variables_expressions: list, with_variables_set: list, variables: list) -> None:
    for counter in range(0, len(variables_expressions)):
        variables[counter] = count_variable_value(variables_expressions[counter], with_variables_set)


def get_accuracy_error(current_iteration: list, previous_iteration: list) -> float:
    error = 0.0
    for counter in range(0, len(current_iteration)):
        error = max(error, abs(current_iteration[counter] - previous_iteration[counter]))
    return error


def solve_by_simple_iterations(main_coefficients: list,
                               free_coefficients: list,
                               epsilon: float,
                               solving_type: SimpleIterationsSolvingType) -> (list, int):
    # выразил переменные из начальных матриц
    variables_expressions = express_main_diagonal_variables(main_coefficients, free_coefficients)

    current_iteration = get_copy(data.INITIAL_VALUES)
    previous_iteration = get_copy(current_iteration)

    # посчитал текущее значение (i + 1)-е с начальными значениями
    count_all_variables_values(variables_expressions, previous_iteration, current_iteration)

    # уже один, так как одна итерация была выше (а была выше перед while, т.к. в Python нет do while)
    iterations_count = 1

    while get_accuracy_error(current_iteration, previous_iteration) > epsilon:
        previous_iteration = get_copy(current_iteration)

        count_all_variables_values(
            variables_expressions,
            previous_iteration if solving_type == SimpleIterationsSolvingType.DEFAULT else current_iteration,
            current_iteration
        )
        iterations_count += 1

    return current_iteration, iterations_count
